In [40]:
import pandas as pd
import numpy as np  
import matplotlib.pyplot as plt
import os
import sys


In [41]:
# load the student data from the CSV file
student_data = pd.read_csv("studentinfo.csv")

In [42]:
student_data.head()

,code_module,code_presentation,id_student,gender,region,highest_education,imd_band,age_band,num_of_prev_attempts,studied_credits,disability,final_result
0,AAA,2013J,11391,M,East Anglian Region,HE Qualification,90-100%,55<=,0,240,N,Pass
1,AAA,2013J,28400,F,Scotland,HE Qualification,20-30%,35-55,0,60,N,Pass
2,AAA,2013J,30268,F,North Western Region,A Level or Equivalent,30-40%,35-55,0,60,Y,Withdrawn
3,AAA,2013J,31604,F,South East Region,A Level or Equivalent,50-60%,35-55,0,60,N,Pass
4,AAA,2013J,32885,F,West Midlands Region,Lower Than A Level,50-60%,0-35,0,60,N,Pass


In [43]:
# Explore the student data
student_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   code_module           32593 non-null  object
 1   code_presentation     32593 non-null  object
 2   id_student            32593 non-null  int64 
 3   gender                32593 non-null  object
 4   region                32593 non-null  object
 5   highest_education     32593 non-null  object
 6   imd_band              31482 non-null  object
 7   age_band              32593 non-null  object
 8   num_of_prev_attempts  32593 non-null  int64 
 9   studied_credits       32593 non-null  int64 
 10  disability            32593 non-null  object
 11  final_result          32593 non-null  object
dtypes: int64(3), object(9)
memory usage: 3.0+ MB


### student_data set having 32593 rows and 12 columns
### id_student & num_of_prev_attempts & studied_credits = 3 columns int data type
### Remaining 9 columns object type


In [44]:
student_data.tail()

,code_module,code_presentation,id_student,gender,region,highest_education,imd_band,age_band,num_of_prev_attempts,studied_credits,disability,final_result
32588,GGG,2014J,2640965,F,Wales,Lower Than A Level,10-20,0-35,0,30,N,Fail
32589,GGG,2014J,2645731,F,East Anglian Region,Lower Than A Level,40-50%,35-55,0,30,N,Distinction
32590,GGG,2014J,2648187,F,South Region,A Level or Equivalent,20-30%,0-35,0,30,Y,Pass
32591,GGG,2014J,2679821,F,South East Region,Lower Than A Level,90-100%,35-55,0,30,N,Withdrawn
32592,GGG,2014J,2684003,F,Yorkshire Region,HE Qualification,50-60%,35-55,0,30,N,Distinction


In [45]:

student_data.describe()

,id_student,num_of_prev_attempts,studied_credits
count,3.259300e+04,32593.000000,32593.000000
mean,7.066877e+05,0.163225,79.758691
std,5.491673e+05,0.479758,41.071900
min,3.733000e+03,0.000000,30.000000
25%,5.085730e+05,0.000000,60.000000
50%,5.903100e+05,0.000000,60.000000
75%,6.444530e+05,0.000000,120.000000
max,2.716795e+06,6.000000,655.000000


In [46]:
student_data.isnull().sum()

code_module                0
code_presentation          0
id_student                 0
gender                     0
region                     0
highest_education          0
imd_band                1111
age_band                   0
num_of_prev_attempts       0
studied_credits            0
disability                 0
final_result               0
dtype: int64

###imd_band column have null values=1111
###need to clean data


In [47]:
student_data.isnull().any()

code_module             False
code_presentation       False
id_student              False
gender                  False
region                  False
highest_education       False
imd_band                 True
age_band                False
num_of_prev_attempts    False
studied_credits         False
disability              False
final_result            False
dtype: bool

In [48]:
###To avoid  rows reduction ,i will replace the null values with zero in the dataset.
### 32592-31482=1110 rows are reduced after removing the null values from the dataset. To avoid this, I will replace the null values with zero in the dataset.

In [49]:
student_data['imd_band'] = student_data['imd_band'].fillna(0)   

In [50]:
student_data.isnull().sum()

code_module             0
code_presentation       0
id_student              0
gender                  0
region                  0
highest_education       0
imd_band                0
age_band                0
num_of_prev_attempts    0
studied_credits         0
disability              0
final_result            0
dtype: int64

In [51]:
student_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   code_module           32593 non-null  object
 1   code_presentation     32593 non-null  object
 2   id_student            32593 non-null  int64 
 3   gender                32593 non-null  object
 4   region                32593 non-null  object
 5   highest_education     32593 non-null  object
 6   imd_band              32593 non-null  object
 7   age_band              32593 non-null  object
 8   num_of_prev_attempts  32593 non-null  int64 
 9   studied_credits       32593 non-null  int64 
 10  disability            32593 non-null  object
 11  final_result          32593 non-null  object
dtypes: int64(3), object(9)
memory usage: 3.0+ MB


### Null values replaced with zeros to avoid data reduction.

In [52]:
### Merge the datasets on the common key or Id.
### Code_module mean code for which module of course.
### cde_presentation is the code for which module of courseis presented.



In [53]:
### Create unique identification for each student by combining the code_module and code_presentation columns.   
student_data['code_module_presentation'] = student_data[['code_module','code_presentation']].apply(lambda x: '_'.join(x), axis=1)   

In [56]:
###Create unique identification for each student by combining the id_student and code_module_presentation columns. 
student_data=student_data.astype({'id_student': str})
student_data['code_module_student_presentation'] = student_data[['id_student','code_module_presentation']].apply(lambda x: '_'.join(x), axis=1)

In [57]:
student_data.head()

,code_module,code_presentation,id_student,gender,region,highest_education,imd_band,age_band,num_of_prev_attempts,studied_credits,disability,final_result,code_module_presentation,code_module_student_presentation
0,AAA,2013J,11391,M,East Anglian Region,HE Qualification,90-100%,55<=,0,240,N,Pass,AAA_2013J,11391_AAA_2013J
1,AAA,2013J,28400,F,Scotland,HE Qualification,20-30%,35-55,0,60,N,Pass,AAA_2013J,28400_AAA_2013J
2,AAA,2013J,30268,F,North Western Region,A Level or Equivalent,30-40%,35-55,0,60,Y,Withdrawn,AAA_2013J,30268_AAA_2013J
3,AAA,2013J,31604,F,South East Region,A Level or Equivalent,50-60%,35-55,0,60,N,Pass,AAA_2013J,31604_AAA_2013J
4,AAA,2013J,32885,F,West Midlands Region,Lower Than A Level,50-60%,0-35,0,60,N,Pass,AAA_2013J,32885_AAA_2013J


In [58]:
student_data.shape

(32593, 14)

In [60]:
uniquevalues = student_data['imd_band'].unique()
print("Unique values in the 'imd_band' column:", uniquevalues)

Unique values in the 'imd_band' column: ['90-100%' '20-30%' '30-40%' '50-60%' '80-90%' '70-80%' 0 '60-70%'
 '40-50%' '10-20' '0-10%']


In [61]:
### Replace the unique values with '10-20' with '10-20%'
student_data['imd_band'] = student_data['imd_band'].replace('10-20', '10-20%')
uniquevalues = student_data['imd_band'].unique()
print("Unique values in the 'imd_band' column after replacement:", uniquevalues)

Unique values in the 'imd_band' column after replacement: ['90-100%' '20-30%' '30-40%' '50-60%' '80-90%' '70-80%' 0 '60-70%'
 '40-50%' '10-20%' '0-10%']


In [62]:
### check duplicate values in the dataset.
duplicates = student_data.duplicated()
print("Number of duplicate rows:", duplicates.sum())

Number of duplicate rows: 0


In [64]:
### checking the unique values for entire dataset.
student_data.nunique()

code_module                             7
code_presentation                       4
id_student                          28785
gender                                  2
region                                 13
highest_education                       5
imd_band                               11
age_band                                3
num_of_prev_attempts                    7
studied_credits                        61
disability                              2
final_result                            4
code_module_presentation               22
code_module_student_presentation    32593
dtype: int64

In [66]:
for col in student_data.columns:
    print(col)
    print(student_data[col].unique())
    print('_' * 40)
    

code_module
['AAA' 'BBB' 'CCC' 'DDD' 'EEE' 'FFF' 'GGG']
________________________________________
code_presentation
['2013J' '2014J' '2013B' '2014B']
________________________________________
id_student
['11391' '28400' '30268' ... '2648187' '2679821' '2684003']
________________________________________
gender
['M' 'F']
________________________________________
region
['East Anglian Region' 'Scotland' 'North Western Region'
 'South East Region' 'West Midlands Region' 'Wales' 'North Region'
 'South Region' 'Ireland' 'South West Region' 'East Midlands Region'
 'Yorkshire Region' 'London Region']
________________________________________
highest_education
['HE Qualification' 'A Level or Equivalent' 'Lower Than A Level'
 'Post Graduate Qualification' 'No Formal quals']
________________________________________
imd_band
['90-100%' '20-30%' '30-40%' '50-60%' '80-90%' '70-80%' 0 '60-70%'
 '40-50%' '10-20%' '0-10%']
________________________________________
age_band
['55<=' '35-55' '0-35']
__________

In [68]:
### Data Cleaning for text columns in the dataset. I will remove the special characters and numbers from the text columns in the dataset.
for col in student_data.select_dtypes(include=['object']).columns:
    student_data[col] = student_data[col].str.strip().str.lower()
    

In [69]:
#inspect the results after cleaning the text columns in the dataset.
for col in student_data.select_dtypes(include=['object']).columns:
    print(col)
    print(student_data[col].unique())
    print('_' * 40)

code_module
['aaa' 'bbb' 'ccc' 'ddd' 'eee' 'fff' 'ggg']
________________________________________
code_presentation
['2013j' '2014j' '2013b' '2014b']
________________________________________
id_student
['11391' '28400' '30268' ... '2648187' '2679821' '2684003']
________________________________________
gender
['m' 'f']
________________________________________
region
['east anglian region' 'scotland' 'north western region'
 'south east region' 'west midlands region' 'wales' 'north region'
 'south region' 'ireland' 'south west region' 'east midlands region'
 'yorkshire region' 'london region']
________________________________________
highest_education
['he qualification' 'a level or equivalent' 'lower than a level'
 'post graduate qualification' 'no formal quals']
________________________________________
imd_band
['90-100%' '20-30%' '30-40%' '50-60%' '80-90%' '70-80%' nan '60-70%'
 '40-50%' '10-20%' '0-10%']
________________________________________
age_band
['55<=' '35-55' '0-35']
________